components/server/src/ome/security/basic/OmeroInterceptor.java

```diff
@@ -91,6 +105,11 @@
 
     private static Logger log = LoggerFactory.getLogger(OmeroInterceptor.class);
 
+    /** thread-safe and may be removed without warning;
+     *  used by {@link omero.cmd.graphs.BaseGraphTraversalProcessor#deleteInstances(String, Collection)} */
+    public static SetMultimap<Session, Long> DELETED_GROUP_LINK_IDS =
+            Multimaps.synchronizedSetMultimap(HashMultimap.<Session, Long>create());
```

Comment:
Me too. Could make private and hide behind a public method that does the `put`s but map would still have to be synchronized. (Could make `preFlush` synchronized instead but that felt even worse.) The architecture of having `Delete2` talk directly to `OmeroInterceptor` which then directly uses `dataSource` is horrifying anyway. It's been tricky to write the interceptor code while working around that most of the methods don't have any sense of which session/transaction they're working in. (Indeed, I'm only using sessions at all here because the current transaction from `Delete2`'s point of view appears not to be the one the interceptor is notified of.)